tietokanta/src/main/resources/db/migration/V1_1094__.sql

```diff
@@ -0,0 +1,164 @@
+-- Lisätään reikäpaikkaukset tietokantaan, reikäpaikkaukset tuodaan Excelistä ja nämä insertoidaan paikkaus tauluun
+CREATE TYPE paikkaustyyppi AS ENUM ('paikkaus', 'reikapaikkaus');
+-- Lisätään sarake paikkaustauluun, annetaan kaikille olemassaoleville tyypiksi 'paikkaus', koska tietokannassa ei ole vielä reikäpaikkauksia
+ALTER TABLE paikkaus ADD COLUMN "paikkaus-tyyppi" paikkaustyyppi DEFAULT 'paikkaus';
+
+-- Sallitaan reikäpaikkauksille paikkauskohde NULL
+ALTER TABLE paikkaus ALTER COLUMN "paikkauskohde-id" DROP NOT NULL;
+
+-- Reikäpaikkaukset ei ole kiinni missään kohteessa, joten meidän tulee sallia NULL arvot
+-- Tehdään uusi constraint, jolla sallitaan NULL jos tyyppi on reikäpaikkaus mutta muuten heitetään virhe
+ALTER TABLE     paikkaus
+ADD CONSTRAINT  paikkauskohde_sallitaanko_null
+CHECK           ("paikkaus-tyyppi" = 'reikapaikkaus' OR "paikkauskohde-id" IS NOT NULL);
+
+-- Lisätään kustannus sarake reikäpaikkauksille
+ALTER TABLE paikkaus ADD COLUMN kustannus NUMERIC;
+
+-- Vaaditaan että kustannus on läsnä reikäpaikkauksille
+ALTER TABLE    paikkaus 
+ADD CONSTRAINT kustannus_sallitaanko_null 
+CHECK          ("paikkaus-tyyppi" = 'paikkaus' OR kustannus IS NOT NULL);
+
+-- Lisätään paikkausmäärä sarake reikäpaikkauksille
+ALTER TABLE paikkaus ADD COLUMN maara INTEGER;
+
+-- Vaaditaan että paikkausmäärä on läsnä reikäpaikkauksille
+ALTER TABLE    paikkaus 
+ADD CONSTRAINT paikkaus_maara_sallitaanko_null 
+CHECK          ("paikkaus-tyyppi" = 'paikkaus' OR maara IS NOT NULL);
+
+-- Lisätään vielä yksikkö sarake reikäpaikkauksille
+ALTER TABLE paikkaus ADD COLUMN "reikapaikkaus-yksikko" VARCHAR(20);
+
+-- Vaaditaan että yksikkö on läsnä reikäpaikkauksille
+ALTER TABLE    paikkaus 
+ADD CONSTRAINT yksikko_sallitaanko_null 
+CHECK          ("paikkaus-tyyppi" = 'paikkaus' OR "reikapaikkaus-yksikko" IS NOT NULL);
+
+-- Jos ulkoinen-id ei ole 0, sen pitää olla jokaisella urakalla uniikki
+-- Eli sallitaan esim arvo 123 urakalle 1 sekä 2, mutta urakalla 1 ei voi olla arvoa 123 kahdesti.
+CREATE UNIQUE INDEX unique_ulkoinen_id_urakalla
+ON paikkaus ("urakka-id", "ulkoinen-id")
+WHERE "ulkoinen-id" <> 0;
+
+
+DROP FUNCTION IF EXISTS reikapaikkaus_upsert(
+  paikkaustyyppi, INT, 
+  TIMESTAMP, INT, 
+  TIMESTAMP, INT, 
+  BOOLEAN, INT, 
+  INT, INT, 
+  TIMESTAMP, TIMESTAMP, 
+  TR_OSOITE, INT, 
+  TEXT, NUMERIC, 
+  NUMERIC, NUMERIC, 
+  NUMERIC, INT, 
+  TEXT, NUMERIC,
+  TEXT, INT,
+  GEOMETRY
+);
+
+
+-- UPSERT funktio, vaikka kaikkia paikkaus sarakkeita ei reikäpaikkauksessa tarvita, lyöty silti mukaan jos tarvitaankin myöhemmin 
+CREATE OR REPLACE FUNCTION reikapaikkaus_upsert(
```

Comment:
Vastustan kevyesti tämänlaista funktiota. Omaan silmään tämän käyttäminen on hämmentävää. Käytännössä kaikissa muissa tilanteissa tämä hoidetaan erillisillä insert/update-lauseilla riippuen siitä onko syötetylle riville annettu id, enkä näe tässäkään tilanteessa syytä, miksi tämä olisi parempi.

Reply:
Tarkoitat ID:llä varmaan ulkoista id:tä? Koska tässä ei nyt muita iditä ole. 
Reikäpaikkauksissa annetaan aina ulkoinen id kun toteumat tuodaan excelistä, ja tämä excel-tuonti olisi tarkoitus yliajaa kaikki olemassa olevat paikkaukset, sekä lisätä puuttuvat

Naputtelin aluksi **INSERT.. CONFLICT x UPDATE ..** 
Mutta **ON CONFLICT** ei toimi tämän unique indexin kanssa, vaikka siihen tuleekin conflikti: 

```sql
CREATE UNIQUE INDEX unique_ulkoinen_id_urakalla
ON paikkaus ("urakka-id", "ulkoinen-id")
WHERE "ulkoinen-id" <> 0;
```

Tämän takia alunperin tuo funktio olemassa. 
Toinen tapa olisi iffitellä tyyliin näin :

```sql
DO $$
BEGIN
	-- Pseudo
	IF EXISTS (	SELECT 1 FROM paikkaus 
				WHERE "urakka-id" = (SELECT id FROM urakka WHERE nimi LIKE 'Muhoksen päällystysurakka')::INT 
				AND "ulkoinen-id" = 1234344::INT )
	THEN
	    -- update
		RAISE NOTICE 'update';
	ELSE
	    -- insert
		RAISE NOTICE 'insert';
	END IF;
END$$;
```

Eli onko ehdotuksia tai mielipidettä, onko tuo iffittely sitten parempi?

Reply:
Puhuttiin jotain että haluttiin clj puolella katsoa onko paikkaus jo olemassa, mutta minulle ei vaan tämä auennut että miksi näin. 

Laitoin tuonne upsertti funktioon tuon selectin, jossa katsotaan onko paikkaus olemassa.
Jos on -> Update
Jos ei ole -> Insert 

> Käytännössä kaikissa muissa tilanteissa tämä hoidetaan erillisillä insert/update-lauseilla

Tekee siis nyt saman asian, mutta funktiona. Tämä on minusta ainakin ihan selkeä keissi. Onko vastustusta, oliko tähän nyt jokin muu ehdotus ja miksi?

https://github.com/finnishtransportagency/harja/pull/3311/commits/b57532a8ccf6588bbc0065f063373c36d543a7eb#diff-fa870ac2eecd29d4f868de32654714fcf432ae3c7f840022d3f3d5f6e92c01b6R80

Reply:
Edelleen olen sitä mieltä, että tämä tarkastus olisi parempi tehdä clj-puolella. `SELECT`in teko insertoidessa tai päivittäessä ei ole normaalia SQL:ää. Ongelma ei siis ole koodin rivimäärässä tai suorituskyvyssä, vaan siinä että tämä rikkoo kaikkia totuttuja käytänteitä. Normaalisti tällainen tilanne käsiteltäisiin harjan koodipesässä tähän tapaan:

```clj
(defn lisaa-tai-paivita-reikapaikkaus [db paikkaus]
  (let [paikkaus-kannassa (q/hae-reikapaikkaus db {:ulkoinen-id (:ulkoinen-id paikkaus)
                                                   :urakka (:urakka paikkaus)})]
    (if (nil? paikkaus-kannassa)
      (q/lisaa-reikapaikkaus! db paikkaus)
      (q/paivita-reikapaikkaus! db paikkaus))))

```

Tällöin insert on insert ja update update. Tämä tarkoittaa yhden ylimääräisen kyselyn tekemistä, joka tarkoittaa pientä suorituskykyhaittaa, mutta se on täysin huomaamaton. Tämä selkeyttää SQL-puolta huomattavasti ja pitää koodipesän yhtenäisenä tämän asian suhteen.